### PR TITLE
 test  case   for native token

### PR DESCRIPTION
**Pull request by tomdong2010**

### CODE REVIEW

File: core/minorblockchain_addon.go [hunk 1]
```diff
@@ -137,6 +137,7 @@ func (m *MinorBlockChain) updateTip(state *state.StateDB, block *types.MinorBloc
 	}
 	if updateTip {
 		m.currentEvmState = state
+		m.confirmedHeaderTip = block.Header()
```
Comment (516108736) [inline core/minorblockchain_addon.go]: not needed
offline to discuss

File: consensus/qkchash/native/native_test.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	b1, _ := shardState.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	assert.Equal(t, len(b1.Transactions()), 1)
+	shardState.FinalizeAndAddBlock(b1)
+	assert.Equal(t, shardState.CurrentHeader(), b1.Header())
+	tTxList, _, err := shardState.GetTransactionByAddress(acc1, &QETH, nil, 20)
+	if err != nil {
+		t.Errorf("GetTransactionByAddress error :%v", err)
+	}
+	assert.NotEqual(t, tTxList[0].Value, &serialize.Uint256{Value: big.NewInt(12345)})
```
Comment (tomdong2010) [inline consensus/qkchash/native/native_test.go]: this is not equal

File: consensus/qkchash/native/native_test.go [hunk 3]
```diff
@@ -97,3 +97,28 @@ func TestDisallowedUnknownToken(t *testing.T) {
...
+	assert.NoError(t, shardState.AddTx(tx1))
+	b1, _ := shardState.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	assert.Equal(t, len(b1.Transactions()), 1)
+	shardState.FinalizeAndAddBlock(b1)
+	assert.Equal(t, shardState.CurrentHeader(), b1.Header())
+	bl, _ := shardState.GetBalance(acc1.Recipient, nil)
+	b2, _ := shardState.GetBalance(acc2.Recipient, nil)
+	bb := new(big.Int).Sub(big.NewInt(10000000), big.NewInt(12345))
+	assert.Equal(t, bl.GetTokenBalance(QETH), new(big.Int).Sub(bb, big.NewInt(21000)))
+	assert.NotEqual(t, b2.GetTokenBalance(QETH), big.NewInt(12345))
```
Comment (tomdong2010) [inline consensus/qkchash/native/native_test.go]: not equal

File: consensus/qkchash/native/native_test.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+	b1 := shardState1.CurrentBlock().CreateBlockToAppend(nil, nil, nil, nil, nil, nil, nil, nil, nil)
+	b1.Header().ParentHash = rootBlock.Header().Hash()
+	val := big.NewInt(888888)
+	gas := new(big.Int).Add(big.NewInt(9000), big.NewInt(21000)).Uint64()
+	tx := core.CreateTransferTransaction(shardState1, id1.GetKey().Bytes(), acc2, acc1, val, &gas, nil, nil, nil, &QKC, &QETHXX)
+	b1.AddTx(tx)
+	deposit := types.CrossShardTransactionDeposit{TxHash: tx.Hash(), From: acc1, To: acc2, Value: &serialize.Uint256{Value: val}, GasPrice: &serialize.Uint256{Value: big.NewInt(1)}, GasTokenID: QKC, TransferTokenID: QETHXX}
+	txL := make([]*types.CrossShardTransactionDeposit, 0)
+	txL = append(txL, &deposit)
+	txList := types.CrossShardTransactionDepositList{txL}
```
Comment (syntrust) [inline consensus/qkchash/native/native_test.go]: use struct key

File: consensus/qkchash/native/native_test.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+	b1.Header().ParentHash = rootBlock.Header().Hash()
+	val := big.NewInt(888888)
+	gas := new(big.Int).Add(big.NewInt(9000), big.NewInt(21000)).Uint64()
+	gasPrice := uint64(2)
+	tx := core.CreateTransferTransaction(shardState1, id1.GetKey().Bytes(), acc2, acc1, val, &gas, &gasPrice, nil, nil, &QETHXX, &QETHXX)
+	b1.AddTx(tx)
+	deposit := types.CrossShardTransactionDeposit{TxHash: tx.Hash(), From: acc1, To: acc2, Value: &serialize.Uint256{Value: val}, GasPrice: &serialize.Uint256{Value: big.NewInt(2)}, GasTokenID: QETHXX, TransferTokenID: QETHXX}
+	txL := make([]*types.CrossShardTransactionDeposit, 0)
+	txL = append(txL, &deposit)
+	txList := types.CrossShardTransactionDepositList{txL}
```
Comment (syntrust) [inline consensus/qkchash/native/native_test.go]: use struct key

File: consensus/qkchash/native/native_test.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
+package native
```
Comment (syntrust) [inline consensus/qkchash/native/native_test.go]: move to core

File: core/native_test.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+	shardState.AddTx(tx)
+	b1, _ := shardState.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	assert.Equal(t, len(b1.Transactions()), 1)
+	evmState, _ := shardState.State()
+	assert.Equal(t, evmState.GetGasUsed(), big.NewInt(0))
+	shardState.FinalizeAndAddBlock(b1)
+	evmState, _ = shardState.State()
+	assert.Equal(t, len(evmState.GetXShardList()), 1)
+	deposit := &types.CrossShardTransactionDeposit{TxHash: tx.Hash(), From: acc1, To: acc2, Value: &serialize.Uint256{Value: val}, GasPrice: &serialize.Uint256{Value: big.NewInt(1)}, GasTokenID: QKC, TransferTokenID: QETHXX}
+	assert.NotEqual(t, evmState.GetXShardList()[0], deposit)
```
Comment (tomdong2010) [inline core/native_test.go]: not equal

File: core/test_utils.go [hunk 8]
```diff
@@ -27,7 +27,7 @@ var (
 	jiaozi                       = new(big.Int).Mul(new(big.Int).SetUint64(1000000000), new(big.Int).SetUint64(1000000000))
 	testShardCoinbaseAmount      = new(big.Int).Mul(new(big.Int).SetUint64(5), jiaozi)
 	testGenesisTokenID           = common.TokenIDEncode("QKC")
-	testGenesisMinorTokenBalance = make(map[string]*big.Int)
+	TestGenesisMinorTokenBalance = make(map[string]*big.Int)
```
Comment (syntrust) [inline core/test_utils.go]: No need to change test_utils.go

File: core/shardstate_test.go [hunk 9]
```diff
@@ -28,7 +28,7 @@ import (
 
 func TestShardStateSimple(t *testing.T) {
 	env := setUp(nil, nil, nil)
-	shardState := createDefaultShardState(env, nil, nil, nil, nil)
+	shardState := CreateDefaultShardState(env, nil, nil, nil, nil)
```
Comment (syntrust) [inline core/shardstate_test.go]: No need to change shardstate_test.go

File: core/shardstate_test.go [hunk 10]
```diff
@@ -2874,3 +2874,216 @@ func TestSigToAddr(t *testing.T) {
...
+	b1, _ := shardState.CreateBlockToMine(nil, &acc2, nil, nil, nil)
+	assert.Equal(t, len(b1.Transactions()), 1)
+	_, _, err := shardState.FinalizeAndAddBlock(b1)
+	if err != nil {
+		t.Logf("smart contract tx is not allowed before evm is enabled ")
+	}
+	assert.Equal(t, shardState.CurrentHeader(), b1.Header())
+	//Check receipts and make sure the transaction is failed
+	//evmState, _ := shardState.State()
+	//state.evm_state.receipts  ?
```
Comment (tomdong2010) [inline core/shardstate_test.go]: state.evm_state.receipts  ?

File: core/native_test.go [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+	assert.Equal(t, tokenBalance2.GetTokenBalance(QETH), big.NewInt(12345))
+	reward := new(big.Int).Add(testShardCoinbaseAmount, big.NewInt(21000)).Uint64()
+	assert.Equal(t, tokenBalance3.GetTokenBalance(QKC), afterTax(reward, shardState))
+	tTxList, _, err := shardState.GetTransactionByAddress(acc1, nil, nil, 0)
+	if err != nil {
+		//t.Errorf("GetTransactionByAddress error :%v", err)
+		t.Log(tTxList)
+	}
+	//not equal
+	assert.NotEqual(t, len(tTxList), 0)
```
Comment (tomdong2010) [inline core/native_test.go]: not equal

File: core/native_test.go [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+	shardState0.AddCrossShardTxListByMinorBlockHash(b1.Header().Hash(), txList)
+	rootBlock = shardState0.GetRootTip().CreateBlockToAppend(nil, nil, nil, nil, nil)
+	rootBlock.AddMinorBlockHeader(shardState0.CurrentBlock().Header())
+	rootBlock.AddMinorBlockHeader(shardState1.CurrentBlock().Header())
+	rootBlock.Finalize(nil, nil, common.EmptyHash)
+	shardState0.AddRootBlock(rootBlock)
+	b2, _ := shardState0.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	shardState0.FinalizeAndAddBlock(b2)
+	balance1, _ := shardState0.GetBalance(acc1.Recipient, nil)
+	assert.Equal(t, balance1.GetTokenBalance(QETHXX), new(big.Int).Add(big.NewInt(999999), big.NewInt(888888)))
```
Comment (tomdong2010) [inline core/native_test.go]: not equal

Comment (tomdong2010) [reply core/native_test.go]: https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L390

File: core/native_test.go (hunk 11 above)
Comment (tomdong2010) [reply core/native_test.go]: https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L81

File: core/native_test.go [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+	tokenBalance, _ := shardState.GetBalance(id1.GetRecipient(), nil)
+	tokenBalance2, _ := shardState.GetBalance(acc2.Recipient, nil)
+	tokenBalance3, _ := shardState.GetBalance(acc2.Recipient, nil)
+	qkcb := tokenBalance.GetTokenBalance(QKC)
+	assert.Equal(t, qkcb, big.NewInt(10000000-21000))
+	assert.Equal(t, tokenBalance.GetTokenBalance(QETH), big.NewInt(99999-12345))
+	assert.Equal(t, tokenBalance2.GetTokenBalance(QETH), big.NewInt(12345))
+	reward := new(big.Int).Add(testShardCoinbaseAmount, big.NewInt(21000)).Uint64()
+	assert.Equal(t, tokenBalance3.GetTokenBalance(QKC), afterTax(reward, shardState))
+	tTxList, _, err := shardState.GetTransactionByAddress(acc1, nil, nil, 0)
```
Comment (tomdong2010) [inline core/native_test.go]: tTxlist is nil  that not true

Comment (tomdong2010) [reply core/native_test.go]: https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L81

File: core/native_test.go [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+	rootBlock.Finalize(nil, nil, common.EmptyHash)
+	shardState0.AddRootBlock(rootBlock)
+	//Add b0 and make sure all x-shard tx's are added
+	b2, _ := shardState0.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	shardState0.FinalizeAndAddBlock(b2)
+	balance1, _ := shardState0.GetBalance(acc1.Recipient, nil)
+	balance3, _ := shardState0.GetBalance(acc1.Recipient, nil)
+	assert.Equal(t, balance1.GetTokenBalance(QETHXX), new(big.Int).Add(big.NewInt(999999), big.NewInt(888888)))
+	reward := new(big.Int).Add(testShardCoinbaseAmount, big.NewInt(18000)).Uint64()
+	assert.Equal(t, balance3.GetTokenBalance(QKC), afterTax(reward, shardState0))
```
Comment (tomdong2010) [inline core/native_test.go]: not equal

Comment (tomdong2010) [reply core/native_test.go]: https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L394

File: core/native_test.go [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+	evmState, _ := shardState0.State()
+	//X-shard gas used
+	assert.Equal(t, evmState.GetGasUsed(), big.NewInt(9000))
```
Comment (tomdong2010) [inline core/native_test.go]: not equal

Comment (tomdong2010) [reply core/native_test.go]: https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L400

File: core/native_test.go [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+	val := big.NewInt(888888)
+	gas := new(big.Int).Add(big.NewInt(9000), big.NewInt(21000)).Uint64()
+	tx := createTransferTransaction(shardState, id1.GetKey().Bytes(), acc1, acc2, val, &gas, nil, nil, nil, &QETHXX, &QETHXX)
+	shardState.AddTx(tx)
+	b1, _ := shardState.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	assert.Equal(t, len(b1.Transactions()), 1)
+	evmState, _ := shardState.State()
+	assert.Equal(t, evmState.GetGasUsed(), 0)
+	shardState.FinalizeAndAddBlock(b1)
+	assert.Equal(t, len(evmState.GetXShardList()), 1)
```
Comment (tomdong2010) [inline core/native_test.go]: not equal

Comment (tomdong2010) [reply core/native_test.go]: https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L447

File: core/native_test.go [hunk 17]
```diff
@@ -226,37 +241,35 @@ func TestXshardNativeTokenReceived(t *testing.T) {
...
 	deposit := types.CrossShardTransactionDeposit{
 		TxHash:          tx.Hash(),
 		From:            acc2,
 		To:              acc1,
 		Value:           &serialize.Uint256{Value: val},
 		GasPrice:        &serialize.Uint256{Value: big.NewInt(2)},
-		GasRemained:     crossShardGas,
-		TransferTokenID: tx.EvmTx.TransferTokenID(),
-		GasTokenID:      tx.EvmTx.GasTokenID(),
+		TransferTokenID:QETHXX,
```
Comment (pingke) [inline core/native_test.go]: format it

**Pull request merged**